mitiq/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,170 @@
+    
+    nqubits = connectivity_graph.number_of_nodes()
+
+    while(len(list(connectivity_graph.edges))>0):
+        num = random.randint(connectivity_graph.size())
+        edges = list(connectivity_graph.edges)
+        curr_edge = edges[num]
+        q1, q2 = curr_edge
+        candidate_edges.add_edge(*curr_edge)
+        connectivity_graph.remove_node(q1)
+        connectivity_graph.remove_node(q2)
+    prob = nqubits*xi / len(list(candidate_edges.edges))
+
+    final_edges = nx.Graph()
+    for edge in list(candidate_edges.edges):
+        if(random.random()<prob):
+            final_edges.add_edge(*edge)
+    return final_edges
+
+
+# %%
+
+
+single_q_cliffords = _single_qubit_cliffords()
+
+
+# %%
+
+
+def random_cliffords(nqubits: int, edges: list) -> cirq.Circuit:
+    """Returns a circuit with a two qubit clifford gate applied to each edge in edges, and a random single qubit
+    clifford gate applied to every other qubit
+
+    Args:
+        nqubits: The number of qubits in the circuit
+        edges: The edges for which the two qubit clifford gate is to be applied
+    """
+    cliffords = single_q_cliffords.c1_in_xy
+
+    gates = [cirq.CNOT(cirq.LineQubit(a), cirq.LineQubit(b)) for a,b in edges]
+
+    for x in range (nqubits):
+        notEdge = True
+        for a,b in edges:
+            if(a==x or b==x):
+                notEdge = False
+        if(notEdge):
+            num = random.randint(24)
+            for clifford_gate in cliffords[num]:
+                gates.append(clifford_gate(cirq.LineQubit(x)))
+    return cirq.Circuit(gates)
+
+
+# %%
+
+
+def random_single_cliffords(nqubits: int) -> cirq.Circuit:
+    """Returns a layer with a random single qubit clifford gate applied on each qubit
+    
+    Args:
+        nqubits: The number of qubits in the circuit
+
+    """
+        
+    circuit = cirq.Circuit()
+    cliffords =  single_q_cliffords.c1_in_xy
+    gates = []
+    for x in range(nqubits):
+        num = random.randint(24)
+        for clifford_gate in cliffords[num]:
+            gates.append(clifford_gate(cirq.LineQubit(x)))
+    circuit.append((gate for gate in gates))
+    return circuit
+
+
+# %%
+
+
+def mirror_circuit(depth: int, xi: float, connectivity_graph: nx.Graph) -> cirq.Circuit:
+    """Returns a randomized mirror circuit
+    
+    Args:
+        depth: The number of random clifford and pauli layers in the mirror circuit including the inverse layers,
+          but excluding the central pauli layer and the initial clifford layer and its inverse
+        xi: Expected two-qubit gate density of the mirror circuit
+        connectivity_graph: The connectivity graph of the backend on which the mirror circuit will be run. This is
+            used to make sure 2 qubit gates are only run on connected qubits
+      """
+    if not 0 <= xi <= 1:
+        raise ValueError("xi should be between 0 and 1")
+    
+    if not depth%4 == 0:
+        raise ValueError("depth must be a multiple of 4")
```

Comment:
It would be annoying to run into this error in practice. You should redefine the input depth parameter to be this depth / 4.